R/utilities-simulation-results.R

```diff
@@ -0,0 +1,132 @@
+#' By default, all individuals from the results are considered. If the individual with the provided ID is not found, the ID is ignored
+#'
+#' @return A list of results, each entry having the name of a quantity path and consisting of the element \code{x} being the time values (in minutes)
+#' and the element \code{y} a \code{NxM} array of simulated values, with \code{N} the number of entries (equals the number of output time steps)
+#' and \code{M} the number of individuals. If no individual contains the results for a given path, the value of the element is \code{NULL}
+#'
+#' @seealso \code{\link{getResultsValues}}
+#'
+#' @export
+#'
+#' @examples
+#' simPath <- system.file("extdata", "simple.pkml", package = "ospsuite")
+#' sim <- loadSimulation(simPath)
+#' simResults <- runSimulation(sim)
+#'
+#' paths <- c(
+#'   "Organism|PeripheralVenousBlood|Caffeine|Plasma (Peripheral Venous Blood)",
+#'   "Organism|ArterialBlood|Plasma|Caffeine|Concentration in container"
+#' )
+#'
+#' resultsValues <- getPopulationResultsValues(simResults, paths, individualIds = 0)
+#' # plot(resultsValues$`Organism|PeripheralVenousBlood|Caffeine|Plasma (Peripheral Venous Blood)`)
+getPopulationResultsValues <- function(simulationResults, quantitiesOrPaths, individualIds = NULL) {
+  validateIsOfType(simulationResults, "SimulationResults")
+  quantitiesOrPaths <- c(quantitiesOrPaths)
+  validateIsOfType(quantitiesOrPaths, c("Quantity", "character"))
+  validateIsNumeric(individualIds, nullAllowed = TRUE)
+
+  # If quantities are passed, get their paths.
+  paths <- quantitiesOrPaths
+  if (isOfType(quantitiesOrPaths, "Quantity")) {
+    paths <- lapply(quantitiesOrPaths, function(x) x$path)
+  }
+  paths <- unique(paths)
+
+  # If no specific individual ids are passed, iterate through all individuals
+  if (is.null(individualIds)) {
+    nrOfIndividuals <- simulationResults$count
+    # TODO: Can we assume that all individuals ids are ascending number from 0 to the number of individuals-1??
+    individualIds <- 0:(nrOfIndividuals - 1)
+  }
+  else {
+    individualIds <- unique(individualIds)
+    nrOfIndividuals <- length(individualIds)
+  }
+
+  # Pre-alloate a list with the number of entries equal to the number of paths
+  output <- vector("list", length(paths))
+  names(output) <- paths
+  # All time values are equal
+  timeVals <- simulationResults$timeValues
+
+  for (path in paths) {
+    # Pre-alloate a list with the number of entries equal to the number of individuals
+    outputValues <- vector("numeric")
+
+    for (individualId in individualIds) {
+      if (simulationResults$hasResultsForIndividual(individualId)) {
+        individualResult <- simulationResults$resultsForIndividual(as.integer(individualId))
+        # Check if the output exist. If not, the entry in the list of outputValues will have the value NULL
+        if (rClr::clrCall(individualResult, "HasValuesFor", path)) {
```

Comment:
Since we do not have an `IndividualResult` object, this code should be moved into our `SimulationResults` . I don't think  we should have any `rClr` stuff leaking outside of our R6 class.